### PR TITLE
Switch to CertEnroll and remove semaphore locking - Faster SSL Generation 

### PR DESCRIPTION
**Pull request by ilushka85**

### CODE REVIEW

**Comment by justcoding121** [comment]
@ilushka85 I can merge this to release. Merge conflicts though.

 I think this also includes changes for proxy authentication. So I would think I can delete the old merges I made to proxy-authentication branch.

**Comment by ilushka85** [comment]
No problem at all.  Can you clean this up with the conflicts?  It does include more proxy authentication / upstream proxy improvements etc.

Just so you understand it now allows fully passing details downstream to various events about which upstream was used and so on.

**Comment by justcoding121** [comment]
@ilushka85 I can clean up the code. Meanwhile #122 need to be fixed before this merge. If you can fix this merge conflict as well.

**Comment by ilushka85** [comment]
Conflicts resolved.  I also fixed the issue from 121.  The important thing to note is that the account must have high enough permission to access certenroll.  For example on my test machine running in a windows service under LocalSystem causes issues.  Under local administrator account it runs fine.  We can leave this as an open bug as it is potentially solvable as the root cert can be generated fine on either account level.

Additionally i started adding support for an external error logging func but it is only implemented in part of the cert area which I will expand on in the future.

File: Titanium.Web.Proxy/Titanium.Web.Proxy.csproj [hunk 1]
```diff
@@ -93,15 +94,20 @@
     <Compile Include="Http\Responses\RedirectResponse.cs" />
     <Compile Include="Shared\ProxyConstants.cs" />
   </ItemGroup>
-  <ItemGroup />
   <ItemGroup>
-    <None Include="app.config" />
-    <None Include="packages.config" />
+    <COMReference Include="CERTENROLLLib">
```
Comment (justcoding121) [inline Titanium.Web.Proxy/Titanium.Web.Proxy.csproj]: Is this COM lib available in default windows installations?

**Comment by ilushka85** [comment]
Yes it's part of windows. All windows internals use this for cert gen.

**Pull request merged**